File: packages/dotnet-sdk/src/TeamsFx/Conversation/NotificationMiddleware.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

namespace Microsoft.TeamsFx.Conversation
{
    using Microsoft.Bot.Builder;
    using Microsoft.Bot.Schema;
    using Microsoft.Bot.Schema.Teams;

    internal class NotificationMiddleware : IMiddleware
    {
        private readonly IConversationReferenceStore _store;

        public NotificationMiddleware(IConversationReferenceStore store)
        {
            _store = store ?? throw new ArgumentNullException(nameof(store));
        }

        public async Task OnTurnAsync(ITurnContext turnContext, NextDelegate next, CancellationToken cancellationToken = default)
        {
            var activityType = ClassifyActivity(turnContext.Activity);
            switch (activityType)
            {
                case ActivityType.CurrentBotInstalled:
                case ActivityType.TeamRestored:
                    {
                        var reference = turnContext.Activity.GetConversationReference();
                        var options = new ConversationReferenceStoreAddOptions {
                            Overwrite = true
                        };
                        await _store.Add(reference.GetKey(), reference, options, cancellationToken).ConfigureAwait(false);
                        break;
                    }
                case ActivityType.CurrentBotMessaged:
                    {
                        await TryAddMessagedReference(turnContext, cancellationToken).ConfigureAwait(false);
                        break;
                    }
                case ActivityType.CurrentBotUninstalled:
                case ActivityType.TeamDeleted:
                    {
                        var reference = turnContext.Activity.GetConversationReference();
                        await _store.Remove(reference.GetKey(), reference, cancellationToken).ConfigureAwait(false);
                        break;
                    }
                default:
                    {
                        break;
                    }
            }

            await next(cancellationToken).ConfigureAwait(false);
        }

        private ActivityType ClassifyActivity(Activity activity)
        {
            var activityType = activity?.Type;
            if (ActivityTypes.InstallationUpdate.Equals(activityType, StringComparison.OrdinalIgnoreCase))
            {
                var action = activity.Action;
                if ("add".Equals(action, StringComparison.OrdinalIgnoreCase))
                {
                    return ActivityType.CurrentBotInstalled;
                }
                else
                {
                    return ActivityType.CurrentBotUninstalled;
                }
            }
            else if (ActivityTypes.ConversationUpdate.Equals(activityType, StringComparison.OrdinalIgnoreCase))
            {
                var eventType = activity.GetChannelData<TeamsChannelData>()?.EventType;
                if ("teamDeleted".Equals(eventType, StringComparison.OrdinalIgnoreCase))
                {
                    return ActivityType.TeamDeleted;
                }
                else if ("teamRestored".Equals(eventType, StringComparison.OrdinalIgnoreCase))
                {
                    return ActivityType.TeamRestored;
                }
            }
            else if (ActivityTypes.Message.Equals(activityType, StringComparison.OrdinalIgnoreCase))
            {
                return ActivityType.CurrentBotMessaged;
            }

            return ActivityType.Unknown;
        }

        private async Task<bool> TryAddMessagedReference(ITurnContext turnContext, CancellationToken cancellationToken = default)
        {
            var reference = turnContext.Activity.GetConversationReference();
            var conversationType = reference?.Conversation?.ConversationType;
            if ("personal".Equals(conversationType, StringComparison.OrdinalIgnoreCase) || "groupChat".Equals(conversationType, StringComparison.OrdinalIgnoreCase))
            {
                var options = new ConversationReferenceStoreAddOptions {
                    Overwrite = false
                };
                var isUpdated = await _store.Add(reference.GetKey(), reference, options, cancellationToken).ConfigureAwait(false);
                return isUpdated;
            }
            else if ("channel".Equals(conversationType, StringComparison.OrdinalIgnoreCase))
            {
                var channelData = turnContext.Activity.GetChannelData<TeamsChannelData>();
                var teamId = channelData?.Team?.Id;
                var channelId = channelData?.Channel?.Id;
                // `teamId == channelId` means General channel. Ignore messaging in non-General channel.
                if (teamId != null && (channelId == null || string.Equals(teamId, channelId)))
                {
                    var channelReference = reference.Clone();
                    channelReference.Conversation.Id = teamId;

                    var options = new ConversationReferenceStoreAddOptions {
                        Overwrite = false
                    };
                    var isUpdated = await _store.Add(channelReference.GetKey(), channelReference, options, cancellationToken).ConfigureAwait(false);
                    return isUpdated;
                }
                else
                {
                    return false;
                }
            }
            else
            {
                return false;
            }
        }
    }
}

File: packages/sdk/src/conversation/middlewares/notificationMiddleware.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { Activity, Middleware, TurnContext } from "botbuilder";
import { cloneConversation, getKey } from "../utils";
import { ConversationReferenceStore } from "../interface";

/**
 * @internal
 */
enum ActivityType {
  CurrentBotInstalled,
  CurrentBotMessaged,
  CurrentBotUninstalled,
  TeamDeleted,
  TeamRestored,
  Unknown,
}

/**
 * @internal
 */
export interface NotificationMiddlewareOptions {
  conversationReferenceStore: ConversationReferenceStore;
}

/**
 * @internal
 */
export class NotificationMiddleware implements Middleware {
  private readonly conversationReferenceStore: ConversationReferenceStore;

  constructor(options: NotificationMiddlewareOptions) {
    this.conversationReferenceStore = options.conversationReferenceStore;
  }

  public async onTurn(context: TurnContext, next: () => Promise<void>): Promise<void> {
    const type = this.classifyActivity(context.activity);
    switch (type) {
      case ActivityType.CurrentBotInstalled:
      case ActivityType.TeamRestored: {
        const reference = TurnContext.getConversationReference(context.activity);
        await this.conversationReferenceStore.add(getKey(reference), reference, {
          overwrite: true,
        });
        break;
      }
      case ActivityType.CurrentBotMessaged: {
        await this.tryAddMessagedReference(context);
        break;
      }
      case ActivityType.CurrentBotUninstalled:
      case ActivityType.TeamDeleted: {
        const reference = TurnContext.getConversationReference(context.activity);
        await this.conversationReferenceStore.remove(getKey(reference), reference);
        break;
      }
      default:
        break;
    }

    await next();
  }

  private classifyActivity(activity: Activity): ActivityType {
    const activityType = activity.type;
    if (activityType === "installationUpdate") {
      const action = activity.action?.toLowerCase();
      if (action === "add" || action === "add-upgrade") {
        return ActivityType.CurrentBotInstalled;
      } else {
        return ActivityType.CurrentBotUninstalled;
      }
    } else if (activityType === "conversationUpdate") {
      const eventType = activity.channelData?.eventType as string;
      if (eventType === "teamDeleted") {
        return ActivityType.TeamDeleted;
      } else if (eventType === "teamRestored") {
        return ActivityType.TeamRestored;
      }
    } else if (activityType === "message") {
      return ActivityType.CurrentBotMessaged;
    }

    return ActivityType.Unknown;
  }

  private async tryAddMessagedReference(context: TurnContext): Promise<void> {
    const reference = TurnContext.getConversationReference(context.activity);
    const conversationType = reference?.conversation?.conversationType;
    if (conversationType === "personal" || conversationType === "groupChat") {
      await this.conversationReferenceStore.add(getKey(reference), reference, { overwrite: false });
    } else if (conversationType === "channel") {
      const teamId = context.activity?.channelData?.team?.id;
      const channelId = context.activity.channelData?.channel?.id;
      // `teamId === channelId` means General channel. Ignore messaging in non-General channel.
      if (teamId !== undefined && (channelId === undefined || teamId === channelId)) {
        const teamReference = cloneConversation(reference);
        teamReference.conversation.id = teamId;
        await this.conversationReferenceStore.add(getKey(teamReference), teamReference, {
          overwrite: false,
        });
      }
    }
  }
}
